Refactor sqlalchemy code in pandas.io.sql to help prepare for sqlalchemy 2.0.

cdcadman opened the pull request:
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

I am splitting this out of #48576 , because it is a major refactor of the code, with the goal of making `SQLDatabase` only accept a `Connection` and not an `Engine`.  sqlalchemy 2.0 restricts the methods that are available to them, which makes it harder to write code that works with both.  For example, `Connection.connect()` creates a branched connection in sqlalchemy 1.x, but is removed in 2.0, but this is called in SQLDatabase.check_case_sensitive().

I also added some clarification on how transactions work in `DataFrame.to_sql`, based on this example, run against pandas 1.5.1:
```
import sqlite3
from pandas import DataFrame
from sqlalchemy import create_engine

with sqlite3.connect(":memory:") as con:
    con.execute("create table test (A integer, B integer)")
    row_count = con.execute("insert into test values (2, 4), (5, 10)").rowcount
    if row_count > 1:
        con.rollback()
    print(con.execute("select count(*) from test").fetchall()[0][0]) # prints 0

with sqlite3.connect(":memory:") as con:
    con.execute("create table test (A integer, B integer)")
    row_count = DataFrame({'A': [2, 5], 'B': [4, 10]}).to_sql('test', con, if_exists='append', index=False)
    if row_count > 1:
        con.rollback() # does nothing, because pandas already committed the transaction.
    print(con.execute("select count(*) from test").fetchall()[0][0]) # prints 2
    
with create_engine("sqlite:///:memory:").connect() as con:
    with con.begin():
        con.exec_driver_sql("create table test (A integer, B integer)")
    try:
        with con.begin():
            row_count = DataFrame({'A': [2, 5], 'B': [4, 10]}).to_sql('test', con, if_exists='append', index=False)
            assert row_count < 2
    except AssertionError:
        pass
    print(con.execute("select count(*) from test").fetchall()[0][0]) # prints 0
```

mroeschke commented:
Thanks for the followup, but could you pair down the changes even more (1 PR for 1 targeted change). It will greatly help the review process. It appears this PR is tackling 3 things which would be better if they were 3 individual PRs

* SQLDatabase accepting connections only
* Disposing engines
* test refactoring

cdcadman commented:
Yes, I think I can do that.  For the first part, where I make SQLDatabase accept Connections only, I might have to disable the tests which pass an Engine, but that will be a much smaller change than the test refactor.

cdcadman commented:
I pulled the engine disposal and test refactoring out of this PR.  I think the engine disposal can wait until that last PR.  But part of my motivation for going with this approach was so that I could put the engine disposal into the `_sqlalchemy_con` function.

mroeschke commented on pandas/io/sql.py:
```diff
@@ -2117,7 +2165,8 @@ def to_sql(
         chunksize=None,
         dtype: DtypeArg | None = None,
         method=None,
-        **kwargs,
```
Could you tackle this change in a separate PR? It looks like this also needs to be fixed in `pandas/core/generic.py` where users first interact with `to_sql`

cdcadman replied:
This change sprung out of making `SQLDatabase` method arguments match `SQLiteDatabase`, which helped me resolve mypy errors.  The `**kwargs` is replaced with the `engine` keyword argument + `**engine_kwargs`, which is what `SQLDatabase.to_sql` expects.  But I could split this out if you prefer.

cdcadman replied:
@mroeschke This change was made to the `SQLDatabase` class as a first step towards implementing #36893.  Currently, `to_sql` is only supported via `sqlalchemy`, but other engines might be added in the future.  So I think it makes sense to not change the signature of `to_sql` in `pandas/core/generic.py`, since end users cannot currently do anything with this functionality.  Let me know what you think.

cdcadman replied:
I split out this change in #49587 .

mroeschke commented on pandas/tests/io/test_sql.py:
```diff
@@ -2525,30 +2525,16 @@ def test_schema_support(self):
 
 
-@pytest.mark.db
-class TestMySQLAlchemy(_TestMySQLAlchemy, _TestSQLAlchemy):
```
So now that connections are always made internally, these tests are redundant?

cdcadman replied:
They no longer test what would actually happen internally, since they pass an Engine to SQLDatabase.  Instead, TestMySQLAlchemyConn tests passing a Connection.

mroeschke commented on pandas/io/sql.py:
```diff
@@ -2066,8 +2114,8 @@ def read_query(
         sql,
         index_col=None,
         coerce_float: bool = True,
-        params=None,
```
Why did the ordering of `params` change?

cdcadman replied:
When I changed the return type of `pandasSQL_builder` to `Iterator[PandasSQL]`, I ran into some code check errors from mypy.  I also tried `Iterator[SQLDatabase] | Iterator[SQLiteDatabase]`.  The way I got mypy checks to pass was by making everything consistent, including argument order, between `SQLDatabase` and `SQLiteDatabase`.

cdcadman replied:
@mroeschke Would it help if I split the following commit out of this PR?  https://github.com/cdcadman/pandas/commit/32fa4cfe23be497ba56af9239cd9f65e03537500

mroeschke replied:
> @mroeschke Would it help if I split the following commit out of this PR? [cdcadman@32fa4cf](https://github.com/cdcadman/pandas/commit/32fa4cfe23be497ba56af9239cd9f65e03537500)

Yes definitely!

cdcadman replied:
I split out this change in https://github.com/pandas-dev/pandas/pull/49587 .

mroeschke commented on pandas/io/sql.py:
```diff
@@ -728,42 +724,67 @@ def has_table(table_name: str, con, schema: str | None = None) -> bool:
+def pandasSQL_builder(
+    con,
+    schema: str | None = None,
+    need_txn: bool = False,
```
```suggestion
    need_transaction: bool = False,
```

IMO it's a bit clearer.

cdcadman replied:
I changed this to `need_transaction`.

Commit:
DOC: Clarify behavior of DataFrame.to_sql

cdcadman commented:
@mroeschke , I've addressed all the comments.  After this PR, the next step would be to refactor the test classes, since the ones involving the sqlalchemy engines would no longer be necessary.  In a later PR, I would dispose of the sqlalchemy engine, if it needed to be created, within the new function `_sqlalchemy_con`.

mroeschke commented on pandas/io/sql.py:
```diff
@@ -729,41 +728,66 @@ def has_table(table_name: str, con, schema: str | None = None) -> bool:
+
+
+@contextmanager
+def _sqlalchemy_con(connectable, need_transaction: bool):
```
Will this be reused in your followup PR? I have a slight preference of just folding this into `pandasSQL_builder` for now until there's a need to share this connection logic

cdcadman replied:
I only plan for `_sqlalchemy_con` to be called by `pandasSQL_builder` and not by anything else.  I split it out to reduce the amount of indentation and keep the functions small.  There is an additional try/finally block and some additional if/else blocks in lines 772-800 here: https://github.com/cdcadman/pandas/blob/sql_fixes/pandas/io/sql.py .  I think I can fold `_sqlalchemy_con` into `pandasSQL_builder` if you still prefer that.

mroeschke replied:
Okay gotcha. I think it'll be fine to leave as a separate function with more code incoming.

However, could you replace `import sqlalchemy` here with `sqlalchemy = import_optional_dependnecy(..., errors="raise")`? Just so that if someone else tries using this function in the future it's known that sqlalchemy is required.

cdcadman replied:
Ok, I made the change.

Commit:
CLN: Make SQLDatabase only accept a sqlalchemy Connection.

mroeschke commented:
Awesome, thanks for the progress here @cdcadman
